web/docs/setup-vast/tune.md

```diff
@@ -170,3 +170,38 @@ be rotated.
 VAST processes log messages in a dedicated thread, which by default buffers up
 to 1M messages for servers, and 100 for clients. The option
 `vast.log-queue-size` controls this setting.
+
+## Rebuilding
+
+The `rebuild` command allows for rebuilding outdated partitions to take
+advantage of improvements by newer VAST versions. This rebuilding process takes
+place in the VAST server in the background.
+
+This process merges partitions up to the configured `max-partition-size`, turns
+VAST v1.x's heterogeneous into VAST v2.x's homogenous partitions, migrates all
+data to the currently configured `store-backend`, and upgrades to the most
+recent internal batch encoding and indexes.
+
+This is how you run it:
+
+```bash
+vast rebuild [--all] [--step-size=<number>] [--parallel=<number] [<expression>]
+```
+-
+The `--all` flag makes the rebuild command consider _all_ partitions rather than
+only outdated ones. This is useful when you change configuration options and
+want to regenerate all partitions.
+
+The `--step-size` and `--parallel` options are performance tuning knobs. The
+step size controls the number of partitions to transform at once, and the
+parallelism level controls how many runs to start in parallel. Both default to 1
+in order to minimize resource usage. Setting the step size to a larger value
+allows for merging undersized partitions.
```

Comment:
I'm struggling with the UX: shouldn't the system figure out the step size? Otherwise we have a correctness issue. The use can accidentally provide step size that's too small and not get the desired effect. For example, if the partition size is currently `S`, and the new size `N` where `N >> S`, then step size `p` must be chosen such that `p * S >= N`. How can the user know this?

Reply:
I discussed this with @tobim and we came up with a few improvements.

First, we changed the catalog lookup to return partitions sorted by their import time range, with newest partitions coming first.

Second, we want to change the rebuild algorithm to look at partitions sorted by their type. This requires a slight change to the code I'll make tomorrow.

Third, we changed the default for the step size to 4. We 4x'd the max partition size for VAST v2.1, so we're hoping this is good enough with the other two changes in mind. I explained what the tradeoff is.

Reply:
> First, we changed the catalog lookup to return partitions sorted by their import time range, with newest partitions coming first.

Woah, I wanted that for a while, but for reasons of deterministic behavior. Assuming there's just one query, the scheduler will not rip apart the order of execution, right? Can I think of this now as plowing through results backwards in time?

> Second, we want to change the rebuild algorithm to look at partitions sorted by their type.

I'm missing context to understand what the effect of this would be.

> Third, we changed the default for the step size to 4. We 4x'd the max partition size for VAST v2.1, so we're hoping this is good enough with the other two changes in mind.

Isn't step size orthogonal to partition size? But yeah, I agree this is probably good enough in practice.

Reply:
> > First, we changed the catalog lookup to return partitions sorted by their import time range, with newest partitions coming first.
> 
> Woah, I wanted that for a while, but for reasons of deterministic behavior. Assuming there's just one query, the scheduler will not rip apart the order of execution, right? Can I think of this now as plowing through results backwards in time?

Not quite currently; this is the relevant snippet from the index:

```cpp
      std::vector<uuid> candidates;
      for (const auto& [_, active_partition] : self->state.active_partitions)
        candidates.push_back(active_partition.id);
      for (const auto& [id, _] : self->state.unpersisted)
        candidates.push_back(id);
      self
        ->request(self->state.catalog, caf::infinite, atom::candidates_v, query)
        .then([=, candidates = std::move(candidates)](
                catalog_result& midx_result) mutable {
          auto& midx_candidates = midx_result.partitions;
          VAST_DEBUG("{} got initial candidates {} and from meta-index {}",
                     *self, candidates, midx_candidates);
          candidates.insert(candidates.end(), midx_candidates.begin(),
                            midx_candidates.end());
          std::sort(candidates.begin(), candidates.end());
          candidates.erase(std::unique(candidates.begin(), candidates.end()),
                           candidates.end());
```

Note that we still sort to be able to deduplicate with `sort`, although there's technically no need to deduplciate that way given how few unpersisted partitions we usually have. The results would also not quite be in most-recently imported order, because the partitions are still scanned from front to back rather than back to front, but I assume this would already be a big improvement.

> > Second, we want to change the rebuild algorithm to look at partitions sorted by their type.
> 
> I'm missing context to understand what the effect of this would be.

This is so the step size of 4 is good enough most of the time. If we take any four partitions we are unlikely to merge anything, so instead we want to take four partitions of the same type that are close in import time range.

Reply:
> Not quite currently; this is the relevant snippet from the index:

I'm not too eager in understanding how to implement this, as I'm lacking context, but I'd like to understand the high-level invariants to get clarity on the mental model. Mind using plain words?

Reply:
We went over this in a call in Slack, but to give a short summary of changes since we started this discussion:

The catalog lookup now returns a vector of partition infos, which contains the type (iff the partition is homogenous) and the number of events. This in turn allowed me to get rid of the step size entirely.

The index now utilizes this to work on partitions by their recency, assuming no partition scheduling is happening. The order is active partitions followed by unpersisted partitions followed by persisted partitions reverse-sorted by their max import time.